docs/proposals/signal-api.md

```diff
@@ -0,0 +1,165 @@
+
+Many applications follow the convention to re-read configuration files when handling the SIGHUP signal. Ability
+to send SIGHUP signal will allow us to reconfigure pods without updating their definition or removing them. Also,
+we can consider using this API in the watch mechanism in kubelet to reconfigure a pod automatically on the ConfigMap
+change.
+
+## Analysis
+
+The system needs to be able to:
+
+1. Model correctly which signals are supported
+2. Allow to define what signals can be used for each pod (and whether they can be used at all)
+3. Provide custom names for the chosen signals in pod
+
+## Proposed design
+
+Signal API will be a part of pod API. There will be no separate API group for that.
+
+#### Schema
+
+The proposed schema is as follow.
+
+There will be a type Notification being a part of the PodSpec, with the following attributes:
+- name
+- type
+- signal
+
+And there will be an API subresource of `pod` called `NotifyREST`, with the options:
+- container
+- name
+
+#### Endpoints
+
+There will be PUT and POST endpoint, rooted at /api/v1/, with the parameters as listed:
+- /namespaces/{namespace}/pods/{pod}/notify
+  - container
+  - name  - name of the notification to send
+
+#### ContainerRuntime interface
+
+The following method will be defined in the ContainerRuntime interface:
+
+```go
+Kill(id, signal string) error
+```
+
+Translating the notification name to the POSIX signal name will happen before using any container runtime, on the
+kubelet API level.
+
+In case of Docker, the `kill` API resource will be called.
+
+In rkt, the only way to signal containers is to use machinectl. Implementing the good way to signal containers in rkt
+seems to be out of scope of this proposal.
+
+## Examples
+
+Let's assume that we have a pod defined like:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: signal-example
+spec:
+  containers:
+    - name: my-container
+      image: myusername/my-program-which-handles-signals
+  notifications:
+    - name: reconfigure
+      type: signal
+      signal: SIGHUP
+    - name: stop
+      type: signal
+      signal: SIGTERM
+```
+
+The contents of the *notifications* section will allow us to do the requests to
+`/api/v1/namespaces/default/pods/signal-example/notify` with data like `{"name": "reconfigure"}` or
+`{"container": "my-container", "name": "reconfigure"}` or `{"name": "stop"}` etc.
+
+And finally, there will be a possibility to do that with kubectl:
+
+```
+kubectl notify signal-example reconfigure
+kubectl notify -c my-container signal-example reconfigure
+kubectl notify --container=my-container signal-example reconfigure
+kubectl notify signal-example stop
+```
+
+## Further improvements
+
+Depending on the further requirements the following features may be added:
+- support for more POSIX signals (i.e. SIGUSR1, SIGUSR2)
+- support for Windows messages
+- keeping notification requests in etcd and send them asynchronously, in case of kube-apiserver or kubelet failover
```

Comment:
Of course it needs discussion. But do we have to have a clear vision of that before merging this spec, or can that be addressed later?